Modules/Scripted/NumpyDataLoader/NumpyDataLoader.py

```diff
@@ -0,0 +1,143 @@
+    """
+        self.parent = parent
+
+
+class NumpyDataLoaderWidget(ScriptedLoadableModuleWidget):
+    def setup(self):
+        ScriptedLoadableModuleWidget.setup(self)
+        # Default reload&test widgets are enough.
+        # Note that reader and writer is not reloaded.
+
+
+class NumpyDataLoaderFileReader(object):
+    def __init__(self, parent):
+        self.parent = parent
+
+    def description(self):
+        return "numpy data file loader"
+
+    def fileType(self):
+        return "numpy"
+
+    def extensions(self):
+        return ["numpy (*.npy)"]
+
+    def canLoadFile(self, filePath):
+        # assume yes if it ends in .npy
+        return True
+
+    def load(self, properties):
+        """
+        uses properties:
+            fileName - path to the .npy file
+        """
+        try:
+            import numpy
+
+            file_path = properties["fileName"]
+
+            # Get node base name from filename
+            if "name" in properties.keys():
+                base_name = properties["name"]
+            else:
+                base_name = os.path.splitext(os.path.basename(file_path))[0]
+            base_name = slicer.mrmlScene.GenerateUniqueName(base_name)
+            scalar_volume = slicer.mrmlScene.AddNewNodeByClass(
+                "vtkMRMLScalarVolumeNode", base_name
+            )
+
+            numpy_image = numpy.load(file_path)
+
+            vtk_image = vtk.vtkImageData()
+            dshape = numpy_image.shape
+            shape3d = dshape[-3:]  # just the last 3 dimensions
+            shape3dr = shape3d[::-1]  # reverse it
+            vtk_image.SetDimensions(shape3dr)
+            vtk_image.AllocateScalars(vtk.VTK_FLOAT, 1)
+            scalar_volume.SetAndObserveImageData(vtk_image)
+
+            node_array = slicer.util.arrayFromVolume(scalar_volume)
+            zero_indices = [0]*(len(dshape) - 3)
+            node_array[:] = numpy_image[zero_indices, :, :, :]
+            slicer.util.arrayFromVolumeModified(scalar_volume)
+
+            scalar_volume.CreateDefaultDisplayNodes()
+
+        except Exception as e:
+            logging.error("Failed to load numpy data file: " + str(e))
+            import traceback
+
+            traceback.print_exc()
+            return False
+
+        self.parent.loadedNodes = [scalar_volume.GetID()]
+        return True
+
+
+class NumpyDataLoaderFileWriter(object):
+    def __init__(self, parent):
+        self.parent = parent
+
+    def description(self):
+        return "numpy data file loader"
+
+    def fileType(self):
+        return "numpy"
+
+    def extensions(self, obj):
+        return ["numpy (*.nii.gz)"]
+
+    def canWriteObject(self, obj):
+        # Only enable this writer in testing mode
+        if not slicer.app.testingEnabled():
+            return False
+
+        return bool(obj.IsA("vtkMRMLScalarVolumeNode"))
+
+    def write(self, properties):
+        return False
+
+
+class NumpyDataLoaderTest(ScriptedLoadableModuleTest):
```

Comment:
It wold be good to have a real test that creates an array, writes it, then confirms it reads back correctly.  Looks like this test will fail.

Reply:
I will add a test in second force-push. In the first, a fix for the rest of the stuff.